File: moderate/valid_parentheses.lua
for line in io.lines(arg[1]) do
  local valid, stack = true, {}
  for i in line:gmatch(".") do
    if i == ")" then
      if #stack > 0 and stack[#stack] == "(" then
        table.remove(stack)
      else
        valid = false
        break
      end
    elseif i == "]" then
      if #stack > 0 and stack[#stack] == "[" then
        table.remove(stack)
      else
        valid = false
        break
      end
    elseif i == "}" then
      if #stack > 0 and stack[#stack] == "{" then
        table.remove(stack)
      else
        valid = false
        break
      end
    else
      stack[#stack+1] = i
    end
  end
  print((valid and #stack == 0) and "True" or "False")
end

File: moderate/valid_parentheses.rb
File.open(ARGV[0]).each_line do |line|
  valid = true
  stack = []
  line.chomp.each_char do |i|
    case i
    when ')'
      if stack.length > 0 && stack[-1] == '(' then stack.delete_at(-1)
      else
        valid = false
        break
      end
    when ']'
      if stack.length > 0 && stack[-1] == '[' then stack.delete_at(-1)
      else
        valid = false
        break
      end
    when '}'
      if stack.length > 0 && stack[-1] == '{' then stack.delete_at(-1)
      else
        valid = false
        break
      end
    else stack << i
    end
  end
  valid &&= stack.empty?
  puts valid ? 'True' : 'False'
end
